internal/config/manager.go

```diff
@@ -0,0 +1,262 @@
+// limitations under the License.
+package config
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/milvus-io/milvus/internal/log"
+	"go.uber.org/zap"
+)
+
+type Manager struct {
+	sync.RWMutex
+	sources      map[string]Source
+	keySourceMap sync.Map
+}
+
+func NewManager() *Manager {
+	return &Manager{
+		sources: make(map[string]Source),
+	}
+}
+
+func (m *Manager) GetConfig(key string) (string, error) {
+	realKey := formatKey(key)
+	sourceName, ok := m.keySourceMap.Load(realKey)
+	if !ok {
+		return "", fmt.Errorf("key not found: %s", key)
+	}
+	return m.configValueBySource(realKey, sourceName.(string))
+}
+
+// Configs returns all the key values
+func (m *Manager) Configs() map[string]string {
+	config := make(map[string]string)
+
+	m.keySourceMap.Range(func(key, value interface{}) bool {
+		sValue, err := m.configValueBySource(key.(string), value.(string))
+		if err != nil {
+			return true
+		}
+		config[key.(string)] = sValue
+		return true
+	})
+
+	return config
+}
+
+func (m *Manager) SetConfig(key, value string) {
+	realKey := formatKey(key)
+	m.sources[CustomSourceName].Set(realKey, value)
+	m.updateEvent(newEvent(CustomSourceName, CreateType, realKey, value))
+}
+
+func (m *Manager) DeleteConfig(key string) {
+	realKey := formatKey(key)
+	for source := range m.sources {
+		m.sources[source].Delete(realKey)
+		m.updateEvent(newEvent(realKey, DeleteType, realKey, ""))
+	}
+}
+
+func (m *Manager) addSource(source Source) error {
+	sourceName := source.GetSourceName()
+	m.Lock()
+	_, ok := m.sources[sourceName]
+	if ok {
+		err := errors.New("duplicate source supplied")
+		m.Unlock()
+		return err
+	}
+
+	m.sources[sourceName] = source
+	m.Unlock()
+
+	err := m.pullSourceConfigs(sourceName)
+	if err != nil {
+		err = fmt.Errorf("failed to load %s cause: %x", sourceName, err)
+		return err
+	}
+
+	return nil
+}
+
+func (m *Manager) pullSourceConfigs(source string) error {
+	m.RLock()
+	configSource, ok := m.sources[source]
+	m.RUnlock()
+	if !ok {
+		return errors.New("invalid source or source not added")
+	}
+
+	config, err := configSource.GetConfigurations()
+	if len(config) == 0 {
+		if err != nil {
+			log.Error("Get configuration by items failed", zap.Error(err))
+			return err
+		}
+		return nil
+	}
+
+	m.updateConfigurationMap(configSource, config)
+	return nil
+}
+
+func (m *Manager) updateConfigurationMap(source Source, configs map[string]string) error {
+	for key := range configs {
+		sourceName, ok := m.keySourceMap.Load(key)
+		if !ok { // if key do not exist then add source
+			m.keySourceMap.Store(key, source.GetSourceName())
+			continue
+		}
+
+		m.RLock()
+		currentSource, ok := m.sources[sourceName.(string)]
+		m.RUnlock()
+		if !ok {
+			m.keySourceMap.Store(key, source.GetSourceName())
+			continue
+		}
+
+		currentSrcPriority := currentSource.GetPriority()
+		if currentSrcPriority > source.GetPriority() { // lesser value has high priority
+			m.keySourceMap.Store(key, source.GetSourceName())
+		}
+	}
```

Comment:
concurrently invoke `updateConfigurationMap`will have some logic issue. An config with lower priority may end up in the result configuration.

Reply:
I can not find the logic issue 😅